File: ocaml/compute_heavy.ml
(* Simulate unsigned 64-bit arithmetic by masking to 64 bits *)
let uint64_mask = 0xFFFFFFFFFFFFFFFFL

let uint64_add a b =
  Int64.logand (Int64.add a b) uint64_mask

let uint64_mul a b =
  Int64.logand (Int64.mul a b) uint64_mask

let matrix_determinant size =
  let n = Int64.to_int (Int64.add size 2L) in
  let matrix = Array.make_matrix n n 0L in

  for i = 0 to n - 1 do
    for j = 0 to n - 1 do
      matrix.(i).(j) <- Int64.rem (Int64.add (Int64.mul (Int64.of_int i) (Int64.of_int j)) 1L) 1000L
    done
  done;

  let det = ref 1L in
  for i = 0 to n - 1 do
    for j = 0 to n - 1 do
      det := Int64.rem (uint64_mul !det matrix.(i).(j)) 1000000L
    done
  done;
  !det

let factorial n =
  if n = 0L || n = 1L then
    1L
  else
    let result = ref 1L in
    for i = 2 to Int64.to_int n do
      result := uint64_mul !result (Int64.of_int i)
    done;
    !result

let hash_compute seed =
  let hash = ref seed in
  for _ = 1 to 100 do
    hash := uint64_add (uint64_mul !hash 1103515245L) 12345L;
    hash := Int64.logxor (Int64.shift_right_logical !hash 16) !hash
  done;
  !hash

let () =
  let start_time = Unix.gettimeofday () in

  let result = ref 0L in
  let iterations = 1_000_000L in

  for i = 0 to Int64.to_int iterations - 1 do
    let i64 = Int64.of_int i in
    result := uint64_add !result (matrix_determinant (Int64.rem i64 15L));
    result := uint64_add !result (factorial (Int64.rem i64 18L));
    result := uint64_add !result (hash_compute i64)
  done;

  let end_time = Unix.gettimeofday () in
  let duration = Int64.of_float ((end_time -. start_time) *. 1_000_000_000.0) in

  Printf.printf "%Ld\n" duration;
  Printf.printf "Result: %Ld\n" !result;
  exit 0

File: lua/compute_heavy.lua
-- Lua implementation with 64-bit integer simulation
local function uint64_mask(n)
    -- Simulate 64-bit unsigned integer overflow
    return n & 0xFFFFFFFFFFFFFFFF
end

local function matrix_determinant(size)
    local n = size + 2
    local matrix = {}

    for i = 0, n - 1 do
        matrix[i] = {}
        for j = 0, n - 1 do
            matrix[i][j] = (i * j + 1) % 1000
        end
    end

    local det = 1
    for i = 0, n - 1 do
        for j = 0, n - 1 do
            det = uint64_mask((det * matrix[i][j]) % 1000000)
        end
    end
    return det
end

local function factorial(n)
    if n == 0 or n == 1 then
        return 1
    end
    local result = 1
    for i = 2, n do
        result = uint64_mask(result * i)
    end
    return result
end

local function hash_compute(seed)
    local hash = seed
    for i = 1, 100 do
        hash = uint64_mask(hash * 1103515245 + 12345)
        hash = uint64_mask((hash >> 16) ~ hash)
    end
    return hash
end

local function get_time_ns()
    local clock = os.clock()
    return math.floor(clock * 1000000000)
end

local start_time = get_time_ns()

local result = 0
local iterations = 1000000

for i = 0, iterations - 1 do
    result = uint64_mask(result + matrix_determinant(i % 15))
    result = uint64_mask(result + factorial(i % 18))
    result = uint64_mask(result + hash_compute(i))
end

local end_time = get_time_ns()
local duration = end_time - start_time

print(duration)
print("Result: " .. result)
